core/src/main/scala/org/apache/spark/scheduler/Stage.scala

```diff
@@ -85,3 +99,8 @@ private[spark] abstract class Stage(
     case _ => false
   }
 }
+
+private[spark] object Stage {
+  // The maximum number of times to retry a stage before aborting 
+  private[scheduler] val maxStageFailures = 4
```

Comment:
@zsxwing this was discussed a decent amount on the [jira](https://issues.apache.org/jira/browse/SPARK-5945).

Personally I still wonder a bit, but we can definitely revisit this later.  This is a huge improvement, so unless you think there is a blocker here, I'd like to get it in as-is for 1.4.